employees/tests/test_unit_report_viewset.py

```diff
@@ -148,3 +151,312 @@ def test_report_detail_view_should_delete_report_on_delete(self):
+        self.url = reverse('custom-report-list')
+
+    def test_custom_list_view_should_display_users_report_list_on_get(self):
+        request = APIRequestFactory().get(path=self.url)
+        request.user = self.user
+        response = ReportList.as_view()(request)
+        self.assertEqual(response.status_code, 200)
+        self.assertContains(response, self.report.description)
+        dictionary = response.data['reports_dict']
+        reports = list(dictionary.values())[0]
+        self.assertTrue(self.report in reports)
+
+    def test_custom_list_view_should_not_be_accessible_for_unauthenticated_user(self):
+        request = APIRequestFactory().get(path=self.url)
+        request.user = AnonymousUser()
+        response = ReportList.as_view()(request)
+        self.assertEqual(response.status_code, 403)
+
+    def test_custom_list_view_should_not_display_other_users_reports(self):
+        other_user = CustomUser(
+            email="otheruser@codepoets.it",
+            password='otheruserpasswd',
+            first_name='Jane',
+            last_name='Doe',
+            country='PL',
+        )
+        other_user.full_clean()
+        other_user.save()
+
+        other_report = Report(
+            date=datetime.datetime.now().date(),
+            description='Some other description',
+            author=other_user,
+            project=self.project,
+            work_hours=Decimal('8.00'),
+        )
+        other_report.full_clean()
+        other_report.save()
+
+        request = APIRequestFactory().get(path=self.url)
+        request.user = self.user
+        response = ReportList.as_view()(request)
+        self.assertEqual(response.status_code, 200)
+        self.assertNotContains(response, other_report.description)
+
+    def test_custom_report_list_view_should_add_new_report_on_post(self):
+        request = APIRequestFactory().post(
+            path=self.url,
+            data={
+                'date': datetime.datetime.now().date(),
+                'description': 'Some description',
+                'project': self.project,
+                'work_hours': Decimal('8.00'),
+            }
+        )
+        request.user = self.user
+        response = ReportList.as_view()(request)
+        self.assertEqual(response.status_code, 201)
+        self.assertEqual(Report.objects.all().count(), 2)
+
+    def test_custom_report_list_view_should_not_add_new_report_on_post_if_form_is_invalid(self):
+        request = APIRequestFactory().post(
+            path=self.url,
+            data={
+                'description': 'Some description',
+                'project': self.project,
+                'work_hours': Decimal('8.00'),
+            }
+        )
+        request.user = self.user
+        response = ReportList.as_view()(request)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(Report.objects.all().count(), 1)
+        self.assertIsNotNone(response.data['errors'])
```

Comment:
It's very tricky to access the error message from here. From what I figured out, it would look like this:
```
self.assertEqual(str(response.data['errors']['date'][0], 'This field is required.')
```
Is this okay?
`response.data['errors']` returns a dictionary *(field_name : list)*.
I use the incorrect field (*'date'*) to access the list of **ErrorDetail** objects.
Then I have to access the first element of this list to get the desired ErrorDetail object.
Since I have no idea how to access string attribute in ErrorDetail, I just use string conversion.

I think it's pretty messy, but that's as far as I got.